data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -158,44 +165,107 @@ private AuthenticationResult acquireAccessTokenUsingDeviceCodeFlow() throws Exce
+                timeout -= 1000;
             }
         }
         return result;
     }
 
     AuthenticationResult acquireWithClientCertificate()
-            throws IOException, InterruptedException, ExecutionException, ServiceUnavailableException{
+            throws InterruptedException, ExecutionException, ServiceUnavailableException {
 
         AuthenticationContext context;
-        AuthenticationResult result;
+        AuthenticationResult result = null;
         ExecutorService service = null;
 
         try {
             service = Executors.newSingleThreadExecutor();
             context = new AuthenticationContext(aadAuthorityUri, false, service);
-            AsymmetricKeyCredential asymmetricKeyCredential = AsymmetricKeyCredential.create(clientCredential.getClientId(),
+            AsymmetricKeyCredential asymmetricKeyCredential = AsymmetricKeyCredential.create(applicationClientId,
                     privateKey, x509Certificate);
             // pass null value for optional callback function and acquire access token
             result = context.acquireToken(clusterUrl, asymmetricKeyCredential, null).get();
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
         } finally {
             if (service != null) {
                 service.shutdown();
             }
         }
+
         if (result == null) {
             throw new ServiceUnavailableException("authentication result was null");
         }
         return result;
     }
 
+    private void acquireToken() throws DataServiceException {
+        lastAuthenticationResultLock.lock();
+        if (lastAuthenticationResult == null || isTokenExpired()) {
+            try {
+                switch (authenticationType) {
+                    case AAD_APPLICATION_KEY:
+                        lastAuthenticationResult = acquireAadApplicationAccessToken();
+                        break;
+                    case AAD_USERNAME_PASSWORD:
+                        lastAuthenticationResult = acquireAadUserAccessToken();
+                        break;
+                    case AAD_DEVICE_LOGIN:
+                        lastAuthenticationResult = acquireAccessTokenUsingDeviceCodeFlow();
+                        break;
+                    case AAD_APPLICATION_CERTIFICATE:
+                        lastAuthenticationResult = acquireWithClientCertificate();
+                        break;
+                    default:
+                        throw new DataServiceException("Authentication type: " + authenticationType.name() + " is invalid");
+                }
+            } catch (Exception e) {
+                throw new DataServiceException(e.getMessage());
+            }
+        }
+        lastAuthenticationResultLock.unlock();
+    }
+
+    private boolean isTokenExpired() {
+        return lastAuthenticationResult.getExpiresOnDate().before(dateInAMinute());
+    }
+
+    AuthenticationResult refreshToken() throws DataServiceException {
```

Comment:
The name is confusing in my opinion - maybe **acquireAccessTokenByRefreshToken** ?